[Bug 18646] Use object handles instead of global object pointers

livecodefraser opened the pull request:
Builds on https://github.com/livecode/livecode/pull/4430

livecodefraser committed:
Add kObjectType constant to all MCObject classes

This value is used to implement safe casting via MCObjectCast.

livecodefraser committed:
Allow MCObjectCast to cast to MCObject*

This is necessary for some template-based magic elsewhere and is always
safe. As such, it doesn't impose any run-time checks.

livecodefraser committed:
Fix MCObjectCast to MCControl*

Audio clips and video clips are MCObject subclasses but are not
MCControls; the previous version allowed them to be cast incorrectly.

livecodefraser committed:
Refactor MCObjectHandle to allow MCButtonHandle, etc

Due to compiler limitations, classes like MCStackHandle won't convert
to MCObjectHandle naturally: this can be done in a type-safe manner
using some features in C++11 that are not available on all of our
supported platforms at this time.

livecodefraser committed:
Update the friend declaration for MCMixinObjectHandle

GCC doesn't like the construction previously used (but Clang does) so
it has been updated to one that is acceptable to both.

livecodefraser committed:
Correct a comment about MCObjectHandles

A "not" was missing, completely reversing the meaning of the comment!

livecodefraser committed:
Merge branch 'develop' of https://github.com/runrev/livecode into mcfoohandle

Conflicts:
	engine/src/aclip.h
	engine/src/button.h
	engine/src/card.h
	engine/src/cpalette.h
	engine/src/eps.h
	engine/src/exec-interface2.cpp
	engine/src/field.h
	engine/src/graphic.h
	engine/src/group.h
	engine/src/image.h
	engine/src/magnify.h
	engine/src/mccontrol.h
	engine/src/object.cpp
	engine/src/object.h
	engine/src/player-legacy.h
	engine/src/player-platform.h
	engine/src/scrolbar.h
	engine/src/stack.h
	engine/src/vclip.h

livecodefraser committed:
Change MCButton::menu to an MCStackHandle

Instead of MCObjectHandle - this removes lots of needless casts.

livecodefraser committed:
Make MCObjectHandle slightly less strict

Get() now no longer asserts that the handle is valid; it instead checks
that it is either unbound or valid. This change has been made to support
a common use pattern in the engine. It should not affect safety as
dereferencing an unbound pointer will cause a segfault.

livecodefraser committed:
Make MCObjectHandle's MCObject* constructor explicit

Having this non-explicit makes various comparisons ambiguous.

livecodefraser committed:
Add comparisons between MC...Handles

livecodefraser committed:
Tweak object.h to not depend on globals.h

This allows globals.h to include object.h (needed in order to turn
various global variables into handles instead of raw pointers).

livecodefraser committed:
object.h no longer includes globals.h

Various other files needed fixed due to implicit dependencies.

livecodefraser committed:
Turn MCObject::parent into an MCObjectHandle

This helps catch any erroneous accesses to parents that have been
deleted.

livecodefraser committed:
Don't use raw object ptrs in the event queue

In case the object gets deleted while there is a queued event.

livecodefraser committed:
Use ObjectHandles for the recent cards list

livecodefraser committed:
Turn MCtopstackptr into an object handle

livecodefraser committed:
Turn MCdefaultstackptr into an object handle

livecodefraser committed:
Turn MCstaticdefaultstackptr into an object handle

livecodefraser committed:
Turn MCmousestackptr into an object handle

livecodefraser committed:
Turn MCclickstackptr into an object handle

livecodefraser committed:
Turn MCfocusedstackptr into an object handle

livecodefraser committed:
Turn MCdynamiccard into an object handle

livecodefraser committed:
Turn MCmenuobjectptr into an object handle

livecodefraser committed:
Turn MCerrorptr into an object handle

livecodefraser committed:
Turn MCerrorlockptr into an object handle

livecodefraser committed:
Turn MCdefaultmenubar into an object handle

livecodefraser committed:
Turn MCmenubar into an object handle

livecodefraser committed:
Turn MCplayers into an object handle

livecodefraser committed:
Turn MCacptr into an object handle

livecodefraser committed:
Turn MCmagimage into an object handle

livecodefraser committed:
Turn MCmagnifier into an object handle

livecodefraser committed:
Turn MCdragsource into an object handle

livecodefraser committed:
Turn MCdragdest into an object handle

livecodefraser committed:
Turn MCactivefield into an object handle

livecodefraser committed:
Turn MCclickfield into an object handle

livecodefraser committed:
Turn MCfoundfield into an object handle

livecodefraser committed:
Turn MCdropfield into an object handle

livecodefraser committed:
Turn MCactiveimage into an object ahndle

livecodefraser committed:
Turn MCeditingimage into an object handle

livecodefraser committed:
Turn MCtooltip into an object handle

livecodefraser committed:
Turn MCmbstackptr into an object handle

livecodefraser committed:
Turn MCdragtargetptr into an object handle

livecodefraser committed:
Turn MCtracestackptr into an object handle

livecodefraser committed:
Turn MCtracedobject into an object handle

livecodefraser committed:
Use object handles for breakpoints and watched vars

livecodefraser committed:
Fix an invalid pointer cast

livecodefraser committed:
Convert the ID cache to use object handles

livecodefraser committed:
Fix a very weird use of operators

...I don't really know what was going on there!

livecodefraser committed:
Fix various Linux-specific compile errors introduced by object handles

livecodefraser committed:
Fix a friend declaration to make GCC happy

Clang seems to be a lot laxer about friend declarations.

livecodefraser committed:
Don't call MCStack::domess on a nil pointer

MCdefaultstack hasn't been initialised at the time the Linux printer
code runs some LCS to parse some text. (I have no idea why this was
working before...). It now uses the template stack instead.

montegoulding commented:
@livecodefraser this is huge! Are you going to patch the object in `MCSelnode` too I had a crash on that the other day there: http://quality.livecode.com/show_bug.cgi?id=18543

montegoulding commented:
@livecodefraser regarding GCC's friend warnings do you have any idea what this is about:

```
../../engine/src/card.h:35:18: warning: ‘MCHccard’ is already a friend of ‘MCCard’ [enabled by default]
     friend class MCHccard;
```

livecodefraser committed:
Refactor the sel object list to use object handles

livecodefraser committed:
Fix an invalid object handle dereference in the breakpoint code

peter-b commented on engine/src/debug.cpp:
```diff
@@ -612,23 +611,25 @@ void MCB_parsewatches(MCExecContext& ctxt, MCStringRef p_input)
+                    t_objecthandle = t_object.object;
+                
+                /* UNCHECKED */ MCNameCreate(*t_hname, &t_handlername);
+                /* UNCHECKED */ MCNameCreate(*t_vname, &t_varname);
```
Please check these! It should be pretty straightforward because you already have a `t_success` local to hand.

peter-b reviewed (approved):
@livecodefraser This is an impressively large patch!  Hopefully it will make a big improvement to stability (once we get release-mode assertions in).

There's a couple of minor things:
- In a couple of places, you've not fixed `/* UNCHECKED */` calls where you probably could have since you were touching the code anyway.
- Where you've reimplemented list structures to use `MCObjectHandle`, you've added cleaning of the list during method calls that probably wouldn't otherwise be expected to modify the list (such as in `MCSellist::redraw()`).  It might have been better to add a `foreach` abstraction which skips over invalid entries, and use it to implement the non-mutating methods that need to only visit valid members.

I think this is okay to merge without addressing either of the points above.

peter-b commented:
@livecode-vulcan review ok 6dbdd74

livecode-vulcan commented:
💙 review by @peter-b ok 6dbdd7445b4203c8ffbc586984059c4df777f018

livecode-vulcan commented:
😞 test failure 6dbdd7445b4203c8ffbc586984059c4df777f018
- try-community-linux-x86: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/438)
- try-community-mac: [pending](https://vulcan.livecode.com/builders/try-community-mac/builds/443)
- try-community-linux-x86_64: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/438)
- try-community-emscripten: [pending](https://vulcan.livecode.com/builders/try-community-emscripten/builds/438)
- try-community-ios: [failure](https://vulcan.livecode.com/builders/try-community-ios/builds/444)
- try-community-win-x86: [pending](https://vulcan.livecode.com/builders/try-community-win-x86/builds/438)
- try-community-android-armv6: [pending](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/438)

peter-b commented:
@livecodefraser As discussed earlier, please rebase this all against the `develop-8.1` branch in order to bring the stability improvements to 8.1.x.  I will file a tracking bug and mark the deletion-pool/object handles bugs as blocked by it.

peter-b commented:
@livecodefraser I have filed [bug 18646](http://quality.livecode.com/show_bug.cgi?id=18646) to track this.

livecodefraser committed:
Fix the mobile-specific event code after MCObjectHandle refactor

The event code for touch events is only compiled on mobile platforms and
was missed in the first refactoring pass.

livecodefraser committed:
Fix Android-specific code broken by MCObject* replacement

livecodefraser committed:
Merge branch 'global-object-handles' of https://github.com/livecodefraser/livecode into global-object-handles

peter-b commented:
@livecode-vulcan review ok 23df397

livecode-vulcan commented:
💙 review by @peter-b ok 23df3970e0216c3ab8546760f1d035e382eef73d

livecodefraser committed:
Fix iOS-specific code broken by MCObject* replacement

peter-b commented:
This is now going into `develop-8.1` via #4804
